torch/onnx/_internal/diagnostics/infra/_infra.py

```diff
@@ -0,0 +1,500 @@
+            [
+                (
+                    formatter._kebab_case_to_snake_case(rule.name),
+                    type(rule),
+                    dataclasses.field(default=rule),
+                )
+                for rule in rules
+            ],
+            bases=(cls,),
+        )()
+
+
+class DiagnosticTool:
+    _sarif_tool: sarif_om.Tool
+    _rules: RuleCollection
+    _triggered_rules: Set[Rule]
+
+    def __init__(
+        self,
+        name: str,
+        version: str,
+        rules: RuleCollection,
+        diagnostic_type: type = Diagnostic,
+    ) -> None:
+        self._sarif_tool = sarif_om.Tool(
+            driver=sarif_om.ToolComponent(name=name, version=version)
+        )
+        self._rules = rules
+        self._triggered_rules = set()
+        if not issubclass(diagnostic_type, Diagnostic):
+            raise TypeError(
+                "Expected diagnostic_type to be a subclass of Diagnostic, "
+                f"but got {diagnostic_type}"
+            )
+        self._diagnostic_type = diagnostic_type
+
+    def sarif(self) -> sarif_om.Tool:
+        """Returns the SARIF Tool object."""
+        self._sarif_tool.driver.rules = [rule.sarif() for rule in self._triggered_rules]
+        return self._sarif_tool
+
+    @property
+    def name(self) -> str:
+        """The name of the tool."""
+        return self._sarif_tool.driver.name
+
+    @property
+    def version(self) -> str:
+        """The version of the tool."""
+        return self._sarif_tool.driver.version
+
+    @property
+    def rules(self) -> RuleCollection:
+        """The rules supported by the tool."""
+        return self._rules
+
+    def create_diagnostic(
+        self,
+        rule: Rule,
+        level: Level,
+        message_args: Optional[Tuple[Any, ...]],
+        **kwargs,
+    ) -> Diagnostic:
+        """Creates a diagnostic for the given arguments.
+
+        Args:
+            rule: The rule that triggered the diagnostic.
+            level: The level of the diagnostic.
+            message_args: The arguments to format the rule's message template.
+            **kwargs: Additional arguments to pass to the Diagnostic constructor.
+
+        Returns:
+            The created diagnostic.
+
+        Raises:
+            ValueError: If the rule is not supported by the tool.
+        """
+        if rule not in self._rules:
+            raise ValueError(
+                f"Rule '{rule.id}:{rule.name}' is not supported by this tool '{self.name} {self.version}'."
+                f" Supported rules are: {self._rules._rule_id_name_set}"
+            )
+        self._triggered_rules.add(rule)
+        return self._diagnostic_type(rule, level, message_args, **kwargs)
+
+
+class Invocation:
+    _sarif_invocation: sarif_om.Invocation
+    # TODO: Implement this.
+
+
+@dataclasses.dataclass
+class DiagnosticOptions:
+    """
+    Options for diagnostic context.
+    """
+
+
+class DiagnosticContext:
```

Comment:
Is this a context for "user" or for "onnx engineers"? can we have both?

Would be possible to have the engineer context to store the current method/function's input *automatically* captured  and stored as metadata?

For a crash on a symbolic, having the inputs from that particular symbolic function being saved for debugging would make debugging a lot easier.

Reply:
imo the context should contain all information possible. When we show it to users they get a view of what is appropriate.

Or maybe we allow different levels of detail?

Reply:
IMO context is bound to one particular export run. Through context api, engineers record diagnostics. Context is used inside the code block that needs diagnoses.

Engine provides the api to create contexts, as well as the api to inspect logs. So engine is used outside of the code block that needs diagnoses. If a users run a sequence of export calls in their script, each call will have a respectful 'run' log recorded in the final log retrievable through engine. 

For lightweight users, perhaps they don't need to interact with any of these. Maybe just an 'options' argument/environment variable to alter the verboseness or the level of details to (capture)display to them.

Reply:
I agree about having one context per run. My suggestion was that, in production, a "lightweight context" is used by default, just to give actionable error messages to the user, very succinct on what is displayed/stored.

Given the repro steps, an onnx engineer could enable the "debug context" and run the model again. On that second run, lots of more information would be presented/stored.

Maybe it is not the case of having different context, but using different `Level`(?) for the same "log"

Reply:
Agree, maybe one of, or both, of diagnostic options and environment variables (convenient for debugging without changing code).